gglasso/solver/single_admm_solver.py

```diff
@@ -7,154 +7,179 @@
+
+    psi = ((p**2 + p)/2)  # number of elements of off-diagonal matrix
+    e_pri = psi * eps_abs + eps_rel * np.maximum(np.linalg.norm(Omega), np.linalg.norm(Theta))
+    e_dual = psi * eps_abs + eps_rel * np.linalg.norm(X)
+
+    r_k = np.linalg.norm(Omega - Theta)
+    s_k = np.linalg.norm(Theta - Theta_t_1)
+
+    term4 = 0
+    if latent:
+        eigD, eigQ = np.linalg.eigh(L - X)
+        proxL = prox_rank_norm(A=L - X, beta=mu1, D=eigD, Q=eigQ)
+        term4 = np.linalg.norm(L - proxL) / (1 + np.linalg.norm(L))
+
+    status_set = set()
+    # primal convergence condition
+    if r_k <= e_pri:
+        status_set.add("primal optimal")
+    # dual convergence condition
+    if s_k <= e_dual:
+        status_set.add("dual optimal")
+
+    stop_value = max(r_k, s_k, term4)
+
+    return stop_value, status_set
+
+
+def ADMM_SGL(S, lambda1, Omega_0, Theta_0=np.array([]), X_0=np.array([]), rho=1., max_iter=1000, tol=1e-7,
+             verbose=False, measure=False, latent=False, mu1=None):
     """
     This is an ADMM algorithm for solving the Single Graphical Lasso problem
-    
+
     Omega_0 : start point -- must be specified as a (p,p) array
     S : empirical covariance matrix -- must be specified as a (p,p) array
-    
+
     latent: boolean to indidate whether low rank term should be estimated
     mu1: low rank penalty paramater, if latent=True
-    
-    
+
+
     In the code, X are the SCALED (with 1/rho) dual variables, for the KKT stop criterion they have to be unscaled (i.e. take rho*X) again!
     """
     assert Omega_0.shape == S.shape
     assert S.shape[0] == S.shape[1]
     assert lambda1 > 0
-        
+
     if latent:
         assert mu1 is not None
         assert mu1 > 0
-        
-    (p,p) = S.shape
-    
+
+    (p, p) = S.shape
+
     assert rho > 0, "ADMM penalization parameter must be positive."
-        
-    
-    # initialize 
-    status = 'not optimal'
+
+    # initialize
+    # status = {0: "not optimal", 1: "primal optimal", 2: "dual optimal"}
+    status = "not optimal"
     Omega_t = Omega_0.copy()
     if len(Theta_0) == 0:
         Theta_0 = Omega_0.copy()
     if len(X_0) == 0:
-        X_0 = np.zeros((p,p))
+        X_0 = np.zeros((p, p))
 
     Theta_t = Theta_0.copy()
-    L_t = np.zeros((p,p))
+    Theta_t_1 = Theta_0.copy()
+    L_t = np.zeros((p, p))
     X_t = X_0.copy()
-       
+    eta_A = (1, set())
+
     runtime = np.zeros(max_iter)
-    kkt_residual = np.zeros(max_iter)
-    
+    residual = np.zeros(max_iter)
+
     for iter_t in np.arange(max_iter):
-        
-        eta_A = ADMM_stopping_criterion(Omega_t, Theta_t, L_t, rho*X_t, S , lambda1, latent, mu1)
-        kkt_residual[iter_t] = eta_A
-        
         if measure:
             start = time.time()
-                
-        if eta_A <= eps_admm:
-            status = 'optimal'
+
+        if iter_t > 0:
+            eta_A = ADMM_stopping_criterion(Omega_t, Theta_t, Theta_t_1, L_t, rho * X_t, S, tol, latent, mu1)
```

Comment:
Just a style preference: `ADMM_stopping_criterion` retruns two values, so I would also assign them to two variables, e.g `eta_A` and `status_set`. Then you don't have to index on `eta_A` later on.